### PR TITLE
DRAFT - Adding ASA and AireOS support

### PR DESCRIPTION
**Pull request by writememe**
# Description

Adding Cisco ASA and Cisco AireOS support to scrapli community drivers.

NOTE: I've put the reasoning for my naming in #86 . As I said, happy to change the naming if you want.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also
 list any relevant details for your test configuration

# Checklist:

- [x] My code follows the style guidelines of this project (no GitHub actions compalints! run `make lint` before
 committing!)
- [x] I have commented my code, pydocstyle and darglint are happy, docstrings are in google docstring format, and all
 docstrings include a summary, args, returns and raises fields (even if N/A)
- [x] I have added tests that prove my fix is effective or that my feature works, if adding "functional" tests please
 note if there are any considerations for the vrnetlab setup
- [x] New and existing unit tests pass locally with my changes

## Test evidence

```python
(venv) ➜  scrapli_community git:(feature/asa-aireos) make lint
python -m isort .
Skipped 3 files
python -m black .
All done! ✨ 🍰 ✨
111 files left unchanged.
python -m pylama .
python -m pydocstyle .
python -m mypy --strict scrapli_community/
Success: no issues found in 92 source files
(venv) ➜  scrapli_community git:(feature/asa-aireos) make test
python -m pytest \
tests/
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.8.10, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /mnt/c/Users/dfjt/Documents/networking/projects/scrapli_community
plugins: pylama-7.7.1, cov-2.12.1
collected 84 items                                                                                                                                                                                        

tests/unit/aethra/atosnt/test_aethra_atosnt.py ...                                                                                                                                                  [  3%]
tests/unit/aruba/test_aruba_aoscx.py ...                                                                                                                                                            [  7%]
tests/unit/cisco/aireos/test_cisco_aireos.py ..                                                                                                                                                     [  9%]
tests/unit/cisco/asa/test_cisco_asa.py ........                                                                                                                                                     [ 19%]
tests/unit/cisco/cbs/test_cisco_cbs.py .......                                                                                                                                                      [ 27%]
tests/unit/edgecore/ecs/test_edgecore_ecs.py ...                                                                                                                                                    [ 30%]
tests/unit/eltex/esr/test_eltex_esr.py .......                                                                                                                                                      [ 39%]
tests/unit/fortinet/wlc/test_fortinet_wlc.py ...                                                                                                                                                    [ 42%]
tests/unit/huawei/vrp/test_huawei_vrp.py .....                                                                                                                                                      [ 48%]
tests/unit/mikrotik/routeros/test_mikrotik_routeros.py .......                                                                                                                                      [ 57%]
tests/unit/nokia/srlinux/test_nokia_srlinux.py ...                                                                                                                                                  [ 60%]
tests/unit/nokia/sros/test_nokia_sros.py ..........                                                                                                                                                 [ 72%]
tests/unit/paloalto/panos/test_paloalto_panos.py ......                                                                                                                                             [ 79%]
tests/unit/ruckus/fastiron/test_ruckus_fastiron.py ........                                                                                                                                         [ 89%]
tests/unit/scrapli/networkdriver/test_scrapli_networkdriver.py ......                                                                                                                               [ 96%]
tests/unit/siemens/roxii/test_siemens_roxii.py ...                                                                                                                                                  [100%]

=========================================================================================== 84 passed in 0.50s ============================================================================================
(venv) ➜  scrapli_community git:(feature/asa-aireos) 
```

### CODE REVIEW

File: scrapli_community/cisco/aireos/cisco_aireos.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+from scrapli_community.cisco.aireos.async_driver import (
+    default_async_on_close,
+    default_async_on_open,
+)
+from scrapli_community.cisco.aireos.sync_driver import default_sync_on_close, default_sync_on_open
+
+DEFAULT_PRIVILEGE_LEVELS = {
+    "exec": (
+        PrivilegeLevel(
+            pattern=r"^\(Cisco Controller\) >$",
```
Comment (carlmontanari) [inline scrapli_community/cisco/aireos/cisco_aireos.py:13]: Is it always `Cisco Controller` or can users set hostnames?

File: tests/unit/cisco/aireos/test_cisco_aireos.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import re
+
+import pytest
+
+from scrapli_community.cisco.aireos.cisco_aireos import DEFAULT_PRIVILEGE_LEVELS
+
+
+# TODO
```
Comment (carlmontanari) [inline tests/unit/cisco/aireos/test_cisco_aireos.py:8]: I think you've done it 😁

File: scrapli_community/cisco/asa/cisco_asa.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+"""scrapli_community.cisco.asa.cisco_asa"""
+
+from scrapli.driver.network.base_driver import PrivilegeLevel
+from scrapli_community.cisco.asa.async_driver import default_async_on_close, default_async_on_open
+from scrapli_community.cisco.asa.sync_driver import default_sync_on_close, default_sync_on_open
+
+DEFAULT_PRIVILEGE_LEVELS = {
+    "exec": (
+        PrivilegeLevel(
+            pattern=r"^[\S]{1,63}>(\s|)$",
```
Comment (carlmontanari) [inline scrapli_community/cisco/asa/cisco_asa.py:10]: `(\s|)$"` <- guessing this is to match maybe space after the `>`? If so, can we re-write -> `\s?$` -- had to pop into regex101 to figure out the empty or thing 🤯

File: scrapli_community/cisco/asa/cisco_asa.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            previous_priv="",
+            deescalate="",
+            escalate="",
+            escalate_auth=False,
+            escalate_prompt="",
+        )
+    ),
+    "privilege_exec": (
+        PrivilegeLevel(
+            pattern=r"^[\S]{1,63}#(\s|)$",
```
Comment (carlmontanari) [inline scrapli_community/cisco/asa/cisco_asa.py:21]: Can this be more explicit? I see there is `\S` here and exec but some specific char classes for config -- can we use the pattern from config? `\S` is easier to read for sure, but ive been trying to keep these as explicit as possible even if it makes the readability a bit shit. It just feels "safer" that way... perhaps all in my head 🤣

File: tests/unit/cisco/asa/test_cisco_asa.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import re
+
+import pytest
+
+from scrapli_community.cisco.asa.cisco_asa import DEFAULT_PRIVILEGE_LEVELS
+
+
+# TODO
```
Comment (carlmontanari) [inline tests/unit/cisco/asa/test_cisco_asa.py:8]: done!

**Review by carlmontanari — CHANGES_REQUESTED** [review]
Looks awesome! I dropped a few comments in there if ya dont mind taking a peak.

Thanks a bunch for the work on this! These will be great additions!!

File: scrapli_community/cisco/aireos/cisco_aireos.py (hunk 1 above)
Comment (writememe) [reply scrapli_community/cisco/aireos/cisco_aireos.py:13]: Hey @carlmontanari, according to this doco (step 3):

https://www.cisco.com/c/en/us/td/docs/wireless/controller/8-5/config-guide/b_cg85/administration_of_cisco_wlc.html

_**
Note: The system prompt can be any alphanumeric string up to 31 characters. You                                                          can change it by entering the config prompt    command.
**_

I'll adjust the regex accordingly

File: tests/unit/cisco/aireos/test_cisco_aireos.py (hunk 2 above)
Comment (writememe) [reply tests/unit/cisco/aireos/test_cisco_aireos.py:8]: Sorry, that's me being a diligent little programmer so I don't forget 🤠

File: scrapli_community/cisco/aireos/cisco_aireos.py (hunk 1 above)
Comment (writememe) [reply scrapli_community/cisco/aireos/cisco_aireos.py]: It's now been adjusted as requested

File: scrapli_community/cisco/asa/cisco_asa.py (hunk 3 above)
Comment (writememe) [reply scrapli_community/cisco/asa/cisco_asa.py]: No worries, have fixed this as well.

File: scrapli_community/cisco/asa/cisco_asa.py (hunk 4 above)
Comment (writememe) [reply scrapli_community/cisco/asa/cisco_asa.py]: Definitely agree, have fixed this as well.

File: tests/unit/cisco/asa/test_cisco_asa.py (hunk 5 above)
Comment (writememe) [reply tests/unit/cisco/asa/test_cisco_asa.py]: Removed!

**Comment by writememe** [comment]
Apologies for the stuff around @carlmontanari , it's now ready to review. I couldn't get this to work for `aireos`, then re-read the doco relating to `auth_bypass` and `on_open`. 

Have come up with a solution, but not 100% sure whether its acceptable.

Below is the demo script which you can use if you have access to an ASA or a WLC:

```python
from scrapli import Scrapli
from scrapli.logging import enable_basic_logging

enable_basic_logging(file=True, level="debug")

asa_device = {
    "host": "<your_asa_ip>",
    "auth_username": "<your_username>",
    "auth_password": "<your_password>",
    "auth_strict_key": False,
    "platform": "cisco_asa",
    "channel_log": True,
}


aireos_device = {
    "host": "<your_wlc_ip>",
    "auth_username": "<your_username>",
    "auth_password": "<your_password>",
    "auth_strict_key": False,
    "platform": "cisco_aireos",
    "channel_log": True,
}

with Scrapli(**asa_device) as asa_conn:
    result = asa_conn.send_command("show version")
    print(result.result)

with Scrapli(**aireos_device) as aireos_conn:
    result = aireos_conn.send_command("show inventory")
    print(result.result)

```

**Comment by carlmontanari** [comment]
Awesome, looking good!

I had a thought regarding the WLC -- sorry I didn't think of this earlier, I forgot they have that weird auth but not normal auth thing going on ... Wonder if you can use the already built in auth methods in the channel so you don't need to do the sleeping and sending of creds yourself. Check [this out](https://github.com/carlmontanari/scrapli/blob/master/scrapli/channel/async_channel.py#L267) (there is a sync flavor as well).

The only downside to doing it with those methods is if/when I ever make this even simpler and turn community into just yaml/json (so it can be used w/ scrapligo as well) then we'd have to figure out some mechanism for doing that (could be a special string that calls this method or something). Buuuuut I can cross that bridge when I get to it 😜 !

So yeah, if you wanna give that a crack cool, if not lemme know and I will merge as is! Either way is good with me!

Thanks for all the work on this!

Carl

**Comment by writememe** [comment]
hey @carlmontanari, I think we just leave it as is for now and if/when you plan on doing a refactor, I can help with testing the refactored drivers. So if you're happy to merge it, that would be great :)

**Comment by carlmontanari** [comment]
> hey @carlmontanari, I think we just leave it as is for now and if/when you plan on doing a refactor, I can help with testing the refactored drivers. So if you're happy to merge it, that would be great :)

Sounds good to me!

**Pull request merged**